include/polycap.h

```diff
@@ -95,10 +98,11 @@ void polycap_rng_free(polycap_rng *rng);
 // construct a new polycap_photon with its initial position, direction, electric field vector and energy
 polycap_photon* polycap_photon_new(
 	polycap_rng *rng,
-	polycap_vector3 start_coords[3],
-	polycap_vector3 start_direction[3],
-	polycap_vector3 start_electric_vector[3],
-	double energy);
+	polycap_vector3 start_coords,
+	polycap_vector3 start_direction,
+	polycap_vector3 start_electric_vector,
+	size_t n_energies,
```

Comment:
Dit is een beetje problematisch IMHO. Het foton zou enkel een start energie moeten hebben, geen ganse array van energieen. Ik snap wel dat dit nuttig is voor het berekenen voor de transmissie efficiëntie, maar voor XMI-MSIM bijvoorbeeld niet want ik verwacht dat `polycap_photon` vrijwel equivalent is aan de fotonen die ik daar gebruik. Kan je dit eventueel oplossen door te werken met een for loop over alle `energies`?

Reply:
Goh, dat gaat, maar dat vertraagd de boel enorm he.
Nu heb ik voor elke foton reflectie direct een volledig spectrum, terwijl als ik een for loop over alle energieën hierboven moet zetten moet ik eigenlijk voor elke energie opnieuw van hetzelfde foton de coordinaten berekenen waarop het zal reflecteren (wat dus energie onafhankelijk is) etc.

Ik ben uiteraard meer voorstander van in XMI-MSIM alle benodigde energieën mee te geven (in zover mogelijk). En anders kunnen we in principe in MSIM gewoon n_energies=1 stellen en *energies is dan een array van 1 element, met de trage loop dan in MSIM zelf?

Reply:
Hmmm... 

In XMI-MSIM zou ik enkel moeten kunnen weten

  1. of een foton met bepaalde start positie, richting en energie door een polycapillair geraakt.
  2. indien het er door geraakt, wat zijn dan de nieuwe positie, en richting?

Dit zou dus moeten berekend worden voor een zeer groot aantal fotonen zowel bij excitatie als bij detector.
Aangezien de start positie, richting, en vaak ook de energie bepaald worden door samplen uit random distributies, zal dit altijd opnieuw moeten berekend worden veronderstel ik en kan er niet echt iets hergebruikt worden uit een vorige berekening?

Excuses als dit een domme vraag is...

Reply:
Zeker geen domme vraag. Dit is het moment waarop we ingrijpende veranderingen aan de code om alles zo optimaal mogelijk te doen lopen moeten bespreken.

Ik vrees echter dat er niet zo heel veel algemene vereenvoudigingen zijn die we kunnen maken. Hoewel:
- Voor PC als focussing optic kunnen we wel voor het volledige excitation spectrum het PC beschrijven als een soort van filter op zowel energie-intensiteit en divergentie (en eventueel later polarisatie verhouding dmv electric field). Dus in feite de source beschrijven voor het PC, kijken hoe dit er na het PC zou uitzien (door gewoon 5k random fotonen door PC te simuleren) en dan beschrijf je een nieuwe bron adhv gemiddelde output uit PC (dus nieuwe source size is zoveel, divergentie zoveel, ...). Dus hier zie ik wel een zeer elegant gebruik van de huidige code, zonder effectief foton per energie te moeten simuleren.

- Bij confocale is het allemaal wat moeilijker uiteraard. Hier moeten we vermoedelijk wel degelijk elk foton apart simuleren, maar is het misschien toch nuttig om direct alle energieën mee te rekenen die potentieel uit het monster/lucht zouden komen. Ik dacht eigenlijk dat dit in MSIM reeds zo was (dat op vanaf elk punt in het monster (incl luchtlaag etc) wordt gekeken wat de volledige spectrale bijdrage ervan in de detector is)

Reply:
Let wel:
>In XMI-MSIM zou ik enkel moeten kunnen weten
>
  >  of een foton met bepaalde start positie, richting en energie door een polycapillair geraakt.
  >  indien het er door geraakt, wat zijn dan de nieuwe positie, en richting?
Dit gebruikt dan gewoon de polycap_photon_launch functie, ipv de ..._get_transmission_efficiencies. Die functie is op zich redelijk snel hoor. De traagheidsfactor bij polycap komt van het feit dat hij effectief doorrekent tot er 5k (of hoeveel dan ook) fotonen doorheen het capillair zijn geraakt. Bij lage transmissie efficiëntie duurt dit uiteraard nogal een tijd, maar een enkel foton simuleren is denk ik kwestie van miliseconden op meeste systemen (heb het niet gebenched ofzo, dus kan er geen definitief getal op zetten, maar alleszins zeer snel). Dus voor duizenden fotonen nog altijd maar een kwestie van seconden.

Reply:
Mja, de `polycap_description_get_transmission_efficiencies` functie is waardeloos in de eigenlijke MC code, enkel `polycap_photon_launch ` zal nuttig zijn daar.

Ik denk ook niet dat de simulatie van de fotonen door de polycapillair echt de boel sterk zal vertragen, en zelf indien dat zo is, is dat geen ramp: het zal zeker mensen niet ervan weerhouden het te gebruiken 😄 

Ik denk dat je gelijk hebt over hoe de excitation channel polycap kan gebruikt worden als filter om de zaken eventueel daar te versnellen. Dat kunnen we best later eens bekijken.

Reply:
By the way: in het geval van variance reduction komen de fotonen met dezelfde richting en start positie de detector polycapillary in: in dat geval zijn al je verschillende energien wel nuttig natuurlijk.

Niet in brute force mode weliswaar, daar kunnen ze vanuit een willekeurige positie de detector polycapillary raken.

Btw: ik zou nadat `polycap_photon_launch` voor 1 foton is doorlopen de transmissie efficiëntie kunnen gebruiken voor de intensity weight factor voor dat foton in variance reduction mode... is dat mogelijk? Of moeten er daarvoor meerdere gesimuleerd worden?

Reply:
Zolang dat foton in variance reduction met exact zelfde start coordinate en direction begint is dat zeker mogelijk! Maar zodra ze een ander capilair van het pc belichten onder een andere hoek heb je een andere uitkomst natuurlijk (andere curvature van capilair zorgt vr andere transmissie eff).

In principe is het zelf mogelijk, denk ik, om een gesimuleerd foton te spiegelen over het centrum van het polycapillair, en dezelfde trans eff te bekomen. Maar zo'n aanpak zou in praktijk mss nogal intensief kunnen zijn om te berekenen. Mss wel een idee voor later, als variance reduction ofzo. Geen idee of het in praktijk echt iets zou opbrengen op vlak van statistiek vs rekentijd.

Versta ik je trouwens juist dat in msim variance reduction alle fotonen onder een zelfde hoek en van uit dezelfde positie de detector binnen komen?

Reply:
In variance reduction mode worden na elke interactie bekeken hoe vanuit de huidige foton positie de detector nog kan bereikt worden door voor elke mogelijke XRF lijn en scattering vanuit de huidige layer (inclusief atmosfeer!) de fractionele fotonen te gaan berekenen. Hiervoor wordt (momenteel) een willekeurige positie op het detector oppervlak gekozen dat zal gebruikt worden voor al deze fractionele fotonen. In geval van een detector polycapillair kan dan gewoon een willekeurige plaats op de polycapillair ingang. Ik veronderstel dat er dus 1 bepaald capillair zal geselecteerd worden, evenals een bepaalde entry position in dat capillair.

Misschien is het beter de energie uit `polycap_photon_new` te halen, en nieuwe `polycap_photon_launch` functies te gaan definiëren die toelaten de energie afhankelijke transmissie efficiëntie voor dat foton te bepalen? De exit positie en richting zijn ook afhankelijk van de energie, correct? In dat geval zouden de getter functies voor deze `polycap_vector3` waarden, ook de energie als argument moeten nemen...

Reply:
>Misschien is het beter de energie uit `polycap_photon_new` te halen, en nieuwe `polycap_photon_launch` functies te gaan definiëren die toelaten de energie afhankelijke transmissie efficiëntie voor dat foton te bepalen?

Ik ben niet helemaal mee hier. Momenteel geeft `polycap_photon_launch` reeds de energie afhankelijke transmissie efficiëntie voor een foton weer (de `photon->weight` array). Dus daarom dat ik in de photon structure de energie mee geef.

>De exit positie en richting zijn ook afhankelijk van de energie, correct?

Nope. Dit is enkel afhankelijk van start positie en direction, en capillary shape natuurlijk. De energie bepaalt enkel of het foton op die exit positie raakt of niet (en welke fractie er raakt). Je moet het zo zien: als een foton botst op een plat vlak met hoek alfa, gaat het reflecteren onder dezelfde hoek alfa, ongeacht de energie. Behalve natuurlijk als alfa te groot is, waardoor het foton geabsorbeerd wordt (wel energie afhankelijk).
Dit is voor 1 enkel foton natuurlijk he. Als je kijkt naar algemene transmissie efficienties doorheen het volledige polycapillair zal je inderdaad zien dat hogere energiën een kleinere focal spot maken etc, doordat de buitenste capillairen een te hoge curvature hebben om die hogere energie fotonen te kunnen doorlaten.

Reply:
Als ik het goed begrijp is het dus wel degelijk mogelijk het grootste deel (energie-onafhankelijk) van de berekening in `polycap_photon_launch` te laten plaatsvinden en het energie afhankelijk deel (transmissie efficiëntie) in een andere functie die dan daarna kan worden uitgevoerd, en die dan de gewenste energien meegeeft?

Reply:
Nee, dat helaas niet :(
De code loopt dus als volgt:
foton komt capilair binnen, botst met de wand op gegeven coordinaat. Daar wordt voor elke energie gekeken welke fractie aan fotonen (ifv energie) door-reflecteert. Dan wordt de volgende reflectie-coordinaat bepaald (energie onafhankelijk, want is gewoon reflectie van die richtingsvector over hoek alfa bij wijze van spreken), waar dan de eerder berekende gewichten worden aangepast adhv hun reflectiviteit op dat punt, volgende coordinaten berekend enzovoort. Op het einde krjig je dan dus 1 exit coordinate (en richting), met een volledig spectrum aan gewichten. Alle gewichten kleiner dan bvb 1e-4 ofzo kan je dan stellen dat ze volledig geabsorbeerd worden.

Als je die processen gaat splitsen kan je inderdaad heel gemakkelijk het exit coordinate en richting bepalen (reflecties als bovenstaande), maar wanneer je dan de transmissie efficientie wil bepalen, zou je dit eigenlijk opnieuw moeten doen, daar je wel degelijk voor elke reflectie de gewichten moet meetellen. Aangezien de totale transmissie efficientie het product is van de opeenvolgende reflecties.
Je zou dit dus wel in twee aparte functies kunnen steken, maar de ene functie zou toch de andere moeten oproepen, dus ik vind het dan nuttiger om het direct allebei tegelijk te doen. De traagste stap is immers al die verschillende reflecties simuleren, kijken welke fractie reflecteert ifv energie is verwaarloosbaar in rekentijd (bij wijze van spreken).

Ik weet niet of het zo veel duidelijker is?

Reply:
Yep thanks voor de uitleg. Best dus `polycap_photon_new` houden zoals het is.